net: tcp: Install a last ack timer in passive close

jukkar opened the pull request:
If we are in a passive close state, then it is possible that the ack we are waiting is lost. So install a timer (default 100 ms) to close the connection if the last ack is not received on time. The timeout is made much shorter than the default ACK timeout (1 sec) because we are just waiting an ACK to our sent FIN and not acking any data.

jukkar committed:
net: tcp: Install a last ack timer in passive close

If we are in a passive close state, then it is possible that
the ack we are waiting is lost or we do not accept the one peer
sent to us because of some earlier out of memory issue.
So install a timer (using by default the FIN timer value) to
close the connection if the last ack is not received on time.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Some discussion about the issue is found here https://github.com/zephyrproject-rtos/zephyr/discussions/67043

pdgendt commented on subsys/net/ip/tcp.c:
```diff
@@ -1857,6 +1859,40 @@ static void tcp_fin_timeout(struct k_work *work)
 	(void)tcp_conn_close(conn, -ETIMEDOUT);
 }
 
+static void tcp_last_ack_timeout(struct k_work *work)
```
Maybe make these `inline`?

jukkar replied:
Most probably the compiler is able to figure out itself what functions to inline or not. There is an old discussion about this in Linux kernel https://www.kernel.org/doc/local/inline.html and relevant links in that doc to LWN.

jukkar commented:
Marking this still as a draft, I am not sure about the implementation of the FIN_WAIT_2 timeout commit.

jukkar committed:
tests: net: tcp: Verify termination at closing state

Make sure the connection is closed by the timer if the final
ack is lost in closing state.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: tcp: Add support for close callback

Add a function callback that is called when the TCP connection
is closed. This is only available if doing network tests.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: tcp: Shorten the test execution

Make the retransmission timer much shorter so that the test
is run faster (about 50% faster test run in qemu_x86).

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: tcp: Add tests for last FIN handling

Make sure the connection is closed by the timer if the final
fin is lost.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

ssharks commented:
The state diagram of TCP is not easy, especially the closing behavior.

A good diagram is shown in: https://users.cs.northwestern.edu/~agupta/cs340/project2/TCPIP_State_Transition_Diagram.pdf

You have to be aware that TCP has two directions and each direction is being closed separately. In the half closed configuration there is still data transfer possible, although Zephyr does not support that at the moment. In the FIN-WAIT state the host is still allowed to send data to the Zephyr socket.
Closing the socket in the FIN-WAIT-2 state is not according to the TCP standard. In FIN2 we are waiting for a FIN packet, not for an ACK packet.

If there was no timeout in the LAST-ACK this is definitely an error. It could be that in the big overhaul of the TCP FIN behavior https://github.com/zephyrproject-rtos/zephyr/pull/61353 this was missed.

jukkar commented:
> Closing the socket in the FIN-WAIT-2 state is not according to the TCP standard. In FIN2 we are waiting for a FIN packet, not for an ACK packet.

Yes, I understand that, and was hesitant to suggest the FIN commit in this PR. It seemed to resolve the issue the reporter was seeing.

For the passive close case, there was no timer so the first commit in this PR seems to work as expected.

As a general note, this has been very tricky to test. I used `siege` and `apachebench` tools to flood the `dumb_http_server_mt` sample but was able to see the passive close ack case only once.
Original reporter in #67155 used burp tool which seem to trigger this issue easily. Unfortunately I was not able to trigger this on zephyr target board when running the burp in Linux. According to reporter, the burp windows version should trigger the problems, but I do not have Windows so could not try this.

jukkar commented:
> In FIN2 we are waiting for a FIN packet, not for an ACK packet.

I missed what this comment refers to, the commits should not wait ACK in FIN2 afaik.

rlubos commented:
> If there was no timeout in the LAST-ACK this is definitely an error. It could be that in the big overhaul of the TCP FIN behavior https://github.com/zephyrproject-rtos/zephyr/pull/61353 this was missed.

> For the passive close case, there was no timer so the first commit in this PR seems to work as expected.

After looking closer into the code, I wouldn't say that there were not timer. Sure there were no direct timer for leaving `TCP_LAST_ACK` state, but that's because timing out that state was controllled by `send_timer`, retransmitting FIN message. In case no ACK was received for **any** of the FIN retransmissions, then the `send_timer` would close the connection indirectly. So it seems to me that the retransmit timer could've been canceled prematurely, which would be wrong. With the direct timer approach it might still work, but we won't retransmit FIN, uless the timeout is set larger than RTO. Plus we ignore the max retransmit parameter in such case.

I guess we'd need to run that windows tool to see what's so special about it that causes the failures.

nikhilbhume-eaton commented:
Yes sure @rlubos , please try this burp community edition [https://portswigger.net/burp/communitydownload] on windows. 
Now, on this PR it seems the issue is not easily reproduceable with burp tool from windows as well.
But on zephyr main branch it gets reproduced within few minutes.

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -2417,6 +2417,19 @@ static enum tcp_state tcp_enter_time_wait(struct tcp *conn)
 	return TCP_TIME_WAIT;
 }
 
+static void tcp_fin_wait_2_timeout(struct k_work *work)
```
I'm not actually sure if we need a specific timeout logic for FIN_WAIT_2 exclusively. The fin timer should always be set when we enter FIN_WAIT_1 state (which is a prerequisite for FIN_WAIT_2), and should not be reset until the final FIN/ACK reply arrives or timeout occurs. It seems to me, that we're simply missing setting up the timer here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L3053
At least it helps for the case when we're stuck in FIN_WAIT_2 with burp tool (we were stuck as the timer was not set when entered FIN_WAIT_1)

rlubos commented:
@nikhilbhume-eaton Thanks, I was able to reproduce the issue easily with Burp community edition and current Zephyr main.
I'm kind of confused though on how to configure this Burp tool to run the test continuously, the intruder tests end after a few iterations and I can't find the option to change it. Can you provide some details on how you configure the test?

jukkar committed:
net: tcp: Reschedule FIN timer when entering FIN state

The FIN timer was not set when we entered the FIN_WAIT_1 state.
This could cause issues if we did not receive proper packets
from peer. With this fix, the connection is always terminated
even if peer does not respond.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>
Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: dumb_http_server: Sleep after accept error

If accept() returns an error, it typically means that the
system is running out of resources. In this case sleep
some time in order the system to cool down a bit.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: dumb_http_server_mt: Sleep after accept error

If accept() returns an error, it typically means that the
system is running out of resources. In this case sleep
some time in order the system to cool down a bit.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Changes since previous version:
* Removed my FIN timer commit as it is not needed. This is replaced by just adding missing FIN timer in proper place. Thanks Robert for this.
* I left the tests as is, as they seem to work and test things we did not test before.
* Fixed an issue in `dumb_http_server_mt` where we exited the handler thread if `accept()` returned an error. This caused weird out of memory issues in the stack as the incoming data was never accepted.
* Added a small delay if the `accept()` fails in `dumb_http_server*` sample apps. This tries to mitigate an issue where the system is running out of resources during accept. If we do not sleep, the system can easily just busy loop if the incoming data flow is high. The sleep tries to mitigate the issue.

@nikhilbhume-eaton please have a try with this PR if you have time.

nikhilbhume-eaton commented:
Yes sure @jukkar , today I will test latest changes and update.

ssharks commented on subsys/net/ip/tcp.c:
```diff
@@ -28,6 +28,8 @@ LOG_MODULE_REGISTER(net_tcp, CONFIG_NET_TCP_LOG_LEVEL);
 
 #define ACK_TIMEOUT_MS CONFIG_NET_TCP_ACK_TIMEOUT
 #define ACK_TIMEOUT K_MSEC(ACK_TIMEOUT_MS)
+#define LAST_ACK_TIMEOUT_MS tcp_fin_timeout_ms
```
Just one question, what is the default value here?

rlubos replied:
Same as for active close, total retransmission timeout, calculated on boot here:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/tcp.c#L4341

ssharks reviewed (commented):
Looks good, just one question

ssharks commented:
> Yes sure @jukkar , today I will test latest changes and update.

If you are able to reproduce the issues, could you try to grab in in a wireshark trace?

pi-ajp commented:
> Removed my FIN timer commit as it is not needed. This is replaced by just adding missing FIN timer in proper place. Thanks Robert for this.

+1 for this change (aka https://github.com/zephyrproject-rtos/zephyr/pull/67067/commits/7a6bf93bc750e8b266d5473634b03b13aca4f9e2).

Prior to finding this PR, I independently tested the following patch for my project which uses  Zephyr 2.7.5 and it has fixed issues with connections stuck in FIN_WAIT_2.  In my project, TCP is operating over a serial PPP interface.  Under high throughput with a small set of available buffers, net_ppp buffer allocation fails resulting in packets being dropped.
```
[01:09:24.370,000] <err> net_ppp: [0x20007658] cannot allocate pkt
[01:09:24.578,000] <err> net_ppp: [0x20007658] cannot allocate new data buffer
```
My hypothesis is that some tx/rx packets related to TCP session close are dropped and result in stuck FIN_WAIT_2 state.

Zephyr 2.7.5 patch:
```
diff --git i/subsys/net/ip/tcp2.c w/subsys/net/ip/tcp2.c
index 8ca4dbce61..917d199072 100644
--- i/subsys/net/ip/tcp2.c
+++ w/subsys/net/ip/tcp2.c
@@ -1939,6 +1939,9 @@ next_state:
 			/* We are closing the connection, send a FIN to peer */
 			if (conn->in_close && conn->send_data_total == 0) {
 				tcp_send_timer_cancel(conn);
+				k_work_reschedule_for_queue(
+					&tcp_work_q, &conn->fin_timer, FIN_TIMEOUT);
+
 				next = TCP_FIN_WAIT_1;
 
 				tcp_out(conn, FIN | ACK);
```

nikhilbhume-eaton commented:
Issue still persists on this PR with latest changes.

[wireshark_log.zip](https://github.com/zephyrproject-rtos/zephyr/files/13842486/wireshark_log.zip)
Here attached zip contains below file which are captured while reproducing the issue.
1. Wireshark log - entire transcations are captured.
2. Screenshot of the tcp states - at the time of device stops responding to http requests.
3. client.py - simple TCP client script, used to reproduce the issue.

I am using dumb_http_server example and Windows environment to run the tools.
Steps to reproduce:
1. Flash the code with net shell config enabled.
2. Start fresh attack from burp tool as shown in above video.
3. Update the ip address and socket number as per your server/device ip address in client.py
4. Run the client.py multiple time -> if you get any TCP error while running ignore it and retrigger the client.py
5. After some time you can see sockets in hold state and will not recover even after long time.

@rlubos suggested me to test by enabling CONFIG_NET_TCP_KEEPALIVE, working on that will update.

ssharks commented:
Hmmm.. I have looked through the wireshark trace and the state seem to coincide with the Wireshark trace. It is strange that the connection with port 59399 seems to be aborted and does not continue after the host sending 3 bytes.

Use filter ip.addr==192.168.1.254&&tcp.port==59399 to find the affected stream.

As the PC only sends 3 bytes here and does not continue, it could block the the dump_http_server as it is only serving 1 connection at a time.
As I see it, with such behavior of the client / server, the observed behavior is fully expected and not a malfunction of the TCP or networking stack.
If you want to rule out this case, the dump_http_server needs to get a timeout on malfunctioning connections. Supporting multiple connections will also help to make this problem less noticeable.

jukkar commented:
> Supporting multiple connections will also help to make this problem less noticeable.

The `dumb_http_server_mt` supports multiple threads (that is why the `_mt` postfix). @nikhilbhume-eaton you might want to try that sample too.

nikhilbhume-eaton commented:
> Issue still persists on this PR with latest changes.
> 
> [wireshark_log.zip](https://github.com/zephyrproject-rtos/zephyr/files/13842486/wireshark_log.zip) Here attached zip contains below file which are captured while reproducing the issue.
> 
> 1. Wireshark log - entire transcations are captured.
> 2. Screenshot of the tcp states - at the time of device stops responding to http requests.
> 3. client.py - simple TCP client script, used to reproduce the issue.
> 
> I am using dumb_http_server example and Windows environment to run the tools. Steps to reproduce:
> 
> 1. Flash the code with net shell config enabled.
> 2. Start fresh attack from burp tool as shown in above video.
> 3. Update the ip address and socket number as per your server/device ip address in client.py
> 4. Run the client.py multiple time -> if you get any TCP error while running ignore it and retrigger the client.py
> 5. After some time you can see sockets in hold state and will not recover even after long time.
> 
> @rlubos suggested me to test by enabling CONFIG_NET_TCP_KEEPALIVE, working on that will update.

@rlubos 
I tested by enabling CONFIG_NET_TCP_KEEPALIVE, on dumb_http_server example and Windows environment. Followed same steps as mentioned above. Sockets are getting hold as shown below even after half hour of idle state[no http request to device]. I think TCP connection should get released/discarded after the TCP timeout.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/113031970/0296a3de-361e-4f7f-858a-1fe112177020)

nikhilbhume-eaton commented:
> Hmmm.. I have looked through the wireshark trace and the state seem to coincide with the Wireshark trace. It is strange that the connection with port 59399 seems to be aborted and does not continue after the host sending 3 bytes.
> 
> Use filter ip.addr==192.168.1.254&&tcp.port==59399 to find the affected stream.
> 
> As the PC only sends 3 bytes here and does not continue, it could block the the dump_http_server as it is only serving 1 connection at a time. As I see it, with such behavior of the client / server, the observed behavior is fully expected and not a malfunction of the TCP or networking stack. If you want to rule out this case, the dump_http_server needs to get a timeout on malfunctioning connections. Supporting multiple connections will also help to make this problem less noticeable.

I will try by installing timout from applicatication layer [dump_http_server]. 
I assumption is socket should get released after timeout if CONFIG_NET_TCP_KEEPALIVE.

> Hmmm.. I have looked through the wireshark trace and the state seem to coincide with the Wireshark trace. It is strange that the connection with port 59399 seems to be aborted and does not continue after the host sending 3 bytes.
> 
> Use filter ip.addr==192.168.1.254&&tcp.port==59399 to find the affected stream.
> 
> As the PC only sends 3 bytes here and does not continue, it could block the the dump_http_server as it is only serving 1 connection at a time. As I see it, with such behavior of the client / server, the observed behavior is fully expected and not a malfunction of the TCP or networking stack. If you want to rule out this case, the dump_http_server needs to get a timeout on malfunctioning connections. Supporting multiple connections will also help to make this problem less noticeable.

Attacker can do anything to malfunction the device. So cannot ignore situation.
Do you mean using timeout in dump_http_server for such connections can be a workaround ? Means HTTP level Keep Alive ?

jukkar commented:
> Attacker can do anything to malfunction the device. So cannot ignore situation.
> Do you mean using timeout in dump_http_server for such connections can be a workaround ? Means HTTP level Keep Alive ?

The dumb_http_server is very simple application that does not do proper connection management, it just listen connections, and returns pre-defined output. So if the client is not sending anything after the connection establishment, then the server will also wait, and you get this situation where the system seems to "hang", although it is just the application that is waiting data that never comes. The burp tool is perhaps not the best one here to test the stack in this case. The HTTP keep-alive would not help here at all.
The network stack itself works now fine after the fixes with the last ack handling are there.

ssharks commented:
For your application you want to close the socket after a certain time of inactivity. This would be a socket timeout, which is by my knowledge not implemented in Zephyr at the moment. Keep alive is the opposite. This will keep the connection active preventing a socket timeout at the other side and to keep NATs along the way from removing the mapping.

The best you can do is to implement a timeout in your application or use the SO_RCVTIMEO on the recv call and handle the return code in your application appropriately to close the socket after a timeout on the recv call.

rlubos commented:
Keep-alive should help if the connection goes dead (peer drops the connection silently). I had impression from our Discord talk that it was the case, but perhaps I was wrong. If that 's not the case, and the peer keeps the connection open, application-level timeout is the only way to go, as others suggest.

jukkar commented:
About the keep-alive, I was referring to HTTP keep-alive https://http.dev/keep-alive, and not TCP keep-alive.

nikhilbhume-eaton commented:
Ok, I got the context.

Its better to test it on fully functional http server. I will still do some more experiments to confirm, stability of our network layer.
So, till that time, may be we can move on from this topic.
